frontend/apps/reader/modules/readerback.lua

```diff
@@ -1,89 +1,193 @@
+-- will ignore those for the same book location):
+-- Called after initial page is set up
+ReaderBack.onReaderReady = ReaderBack._onViewPossiblyUpdated
+-- New page on paged media or crengine in page mode
+ReaderBack.onPageUpdate = ReaderBack._onViewPossiblyUpdated
+-- New page on crengine in scroll mode
+ReaderBack.onPosUpdate = ReaderBack._onViewPossiblyUpdated
+-- View updated (possibly on the same page) on paged media
+ReaderBack.onViewRecalculate = ReaderBack._onViewPossiblyUpdated
+-- View updated (possibly on the same page) on paged media (needed in Reflow mode)
+ReaderBack.onPagePositionUpdated = ReaderBack._onViewPossiblyUpdated
 
-        if saved_location then
-            ignore_location = self:_getCurrentLocation()
-            logger.dbg("[ReaderBack] restoring:", saved_location)
-            self.ui:handleEvent(Event:new('RestoreBookLocation', saved_location))
+function ReaderBack:onBack()
+    local back_in_reader = G_reader_settings:readSetting("back_in_reader") or "default"
+    local back_to_exit = G_reader_settings:readSetting("back_to_exit") or "prompt"
+
+    if back_in_reader == "previous_read_page" then
+        if #self.location_stack > 0 then
+            local saved_location = table.remove(self.location_stack)
+            if saved_location then
+                -- Reset self.cur_location, which will be updated with the restored
+                -- saved_location, which will then not be added to the stack
+                self.cur_location = nil
+                logger.dbg("[ReaderBack] restoring:", saved_location)
+                self.ui:handleEvent(Event:new('RestoreBookLocation', saved_location))
+                -- Ensure we always have self.cur_location updated, as in some
+                -- cases (same page), no event that we handle might be sent.
+                UIManager:nextTick(self._addPreviousLocationToStackCallback)
+                return true
+            end
+        elseif not self.back_resist or back_to_exit == "disable" then
+            -- Show a one time notification when location stack is empty.
+            -- On next "Back" only, proceed with the default behaviour (unless
+            -- it's disabled, in which case we always show this notification)
+            self.back_resist = true
+            UIManager:show(Notification:new{
+                text = _("Location history is empty."),
+                timeout = 2,
+            })
             return true
+        else
+            self.back_resist = nil
         end
-    else
-        logger.dbg("[ReaderBack] no location history, closing")
+    elseif back_in_reader == "previous_location" then
+        -- ReaderLink maintain its own location_stack of less frequent jumps
+        -- (links or TOC entries followed, skim document...)
+        if back_to_exit == "disable" then
+            -- Let ReaderLink always show its notification if empty
+            self.ui.link:onGoBackLink(true) -- show_notification_if_empty=true
```

Comment:
Ooh, good idea of the Python-like named argument comment, makes those kind of tricks much more readable ;).